### PR TITLE
[build, test] [sagemaker] Scheduler for PR tests as an optional path

### PR DESCRIPTION
**Pull request by Huahan98**
*Issue #, if available:*

## Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [build] | [test] | [build, test] | [ec2, ecs, eks, sagemaker]
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [x] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).

*Description:* 
Enable an optional path to run CI tests through the scheduler, by setting USE_SCHEDULER to True at src/config/test_config.py

*Tests run:*  
- Units test for the scheduler package included under Scheduler/tests 
- integration tests ran through the scheduler

*DLC image/dockerfile:* N/A

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/config/test_config.py [hunk 1]
```diff
@@ -4,3 +4,4 @@
 DISABLE_EKS_TESTS = False
 DISABLE_EC2_TESTS = False
 DISABLE_SANITY_TESTS = False
+USE_SCHEDULER = False
```
Comment (saimidu) [inline src/config/test_config.py:7]: Set this to True so that the results of SageMaker tests on the scheduler can be observed.

Also disable all the other test types apart from SageMaker tests.

File: .gitignore [hunk 2]
```diff
@@ -2,3 +2,8 @@ __pycache__
 logs
 venv
 .idea
+.DS_Store
+*.pyc
+Scheduler/trigger.py
+Scheduler/buildspec.yml
+debug.py
```
Comment (saimidu) [inline .gitignore:9]: Do these need to be here?

Comment (Huahan98) [reply .gitignore]: removed

File: Scheduler/tests/query_and_cancel_test.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+- Call cancel_request on tickets on the request queue, check that the request tickets are removed. 
+- Clean up the artifacts. 
+"""
+
+LOGGER = logging.getLogger(__name__)
+LOGGER.setLevel(logging.DEBUG)
+LOGGER.addHandler(logging.StreamHandler(sys.stdout))
+
+# test parameters
+TEST_ECR_URI = "754106851545.dkr.ecr.us-west-2.amazonaws.com/pr-tensorflow-training:2.2.0-gpu-py37-cu101-ubuntu18.04-example-pr-269-2020-06-11-22-13-27"
```
Comment (saimidu) [inline Scheduler/tests/query_and_cancel_test.py:27]: Change this to use an image from one of the already released images:

https://github.com/aws/deep-learning-containers/blob/master/available_images.md#general-framework-containers

Use the TensorFlow 2.2.0 training image in the table above:
```
763104351884.dkr.ecr.us-east-1.amazonaws.com/tensorflow-training:2.2.0-gpu-py37-cu101-ubuntu18.04
```

File: test/testrunner.py [hunk 4]
```diff
@@ -38,21 +43,137 @@ def run_sagemaker_local_tests(images):
...
+
+
 def run_sagemaker_remote_tests(images):
     """
     Function to set up multiprocessing for SageMaker tests
     :param images: <list> List of all images to be used in SageMaker tests
     """
-    if not images:
+    use_scheduler = os.getenv("USE_SCHEDULER", "False")
+    executor_mode = os.getenv("EXECUTOR_MODE", "False")
```
Comment (saimidu) [inline test/testrunner.py:140]: nit: You can make these variables boolean by comparing the values you get for the env variables to "true".

So it would look like:
```
use_scheduler = os.getenv("USE_SCHEDULER", "False").lower() == "true"
executor_mode = os.getenv("EXECUTOR_MODE", "False").lower() == "true"
```
This way, when you need  to use these variables in if-else conditions, you don't have to compare them to "true" there every time. Especially since these variables are supposed to be representing boolean True/False anyway.

File: test/testrunner.py [hunk 5]
```diff
@@ -38,21 +43,137 @@ def run_sagemaker_local_tests(images):
...
+        if test_succeeded:
+            log_return.update_pool("completed", instance_type, num_of_instances, job_type, test_report)
+        else:
+            log_return.update_pool("runtimeError", instance_type, num_of_instances, job_type, test_report)
         return
-    pool_number = len(images)
-    with Pool(pool_number) as p:
-        p.map(sm_utils.execute_sagemaker_remote_tests, images)
+
+    if use_scheduler.lower() == "true":
```
Comment (saimidu) [inline test/testrunner.py:162]: Should we explicitly prevent executor_mode and use_scheduler from being True at the same time?

If the `return` in line 160 is removed, then it would be possible to send the scheduling into an infinite loop, right?

File: test/testrunner.py [hunk 6]
```diff
@@ -82,16 +204,21 @@ def setup_eks_cluster(framework_name):
...
         resources_location = os.path.join(test_path, "tensorflow", "training", "resources")
         setup_sm_benchmark_tf_train_env(resources_location, tf1_images_in_list, tf2_images_in_list)
 
 
 def main():
     # Define constants
+    start_time = datetime.now()
     test_type = os.getenv("TEST_TYPE")
-    dlc_images = get_dlc_images()
+    executor_mode = os.getenv("EXECUTOR_MODE", "False")
```
Comment (saimidu) [inline test/testrunner.py:217]: Check if `os.getenv(name, "False").lower() == "true"` here, assign to `executor_mode`, and directly use `if executor_mode:` elsewhere.

File: test/testrunner.py [hunk 7]
```diff
@@ -82,16 +204,21 @@ def setup_eks_cluster(framework_name):
...
+    if executor_mode.lower() == "true":
+        dlc_images = os.getenv("DLC_IMAGE")
+    else:
+        dlc_images = get_dlc_images()
```
Comment (saimidu) [inline test/testrunner.py:221]: Can use ternary-if

File: test/testrunner.py [hunk 8]
```diff
@@ -163,17 +291,21 @@ def main():
...
     elif specific_test_type == "sagemaker-local":
         run_sagemaker_local_tests(
             [image for image in standard_images_list if not ("tensorflow-inference" in image and "py2" in image)]
         )
     else:
-        raise NotImplementedError(f"{test_type} test is not supported. "
-                                  f"Only support ec2, ecs, eks, sagemaker and sanity currently")
+        raise NotImplementedError(
+            f"{test_type} test is not supported. " f"Only support ec2, ecs, eks, sagemaker and sanity currently"
+        )
```
Comment (saimidu) [inline test/testrunner.py:308]: I'm guessing this was done by `black`. In that case, you can actually join the two strings here, and replace the `<end_of_str1> " f"<start_of_str2>` with `<end_of_str1> <start_of_str2>`

Comment (Huahan98) [reply test/testrunner.py:333]: fixed

File: test/testrunner.py (hunk 6 above)
Comment (Huahan98) [reply test/testrunner.py:232]: fixed

File: Scheduler/tests/jobrequester_test.py [hunk 9]
```diff
@@ -17,7 +15,7 @@
 LOGGER.setLevel(logging.DEBUG)
 LOGGER.addHandler(logging.StreamHandler(sys.stdout))
 
-TEST_IMAGE = "754106851545.dkr.ecr.us-west-2.amazonaws.com/pr-tensorflow-training:2.2.0-gpu-py37-cu101-ubuntu18.04-example-pr-269-2020-06-11-22-13-27"
+TEST_IMAGE = "763104351884.dkr.ecr.us-east-1.amazonaws.com/tensorflow-training:2.2.0-gpu-py37-cu101-ubuntu18.04"
```
Comment (saimidu) [inline Scheduler/tests/jobrequester_test.py:18]: Tests run in us-west-2 will fail if you use the image from us-east-1. Set the region to be "us-west-2".

File: Scheduler/tests/query_and_cancel_test.py [hunk 10]
```diff
@@ -24,7 +24,7 @@
 LOGGER.addHandler(logging.StreamHandler(sys.stdout))
 
 # test parameters
-TEST_ECR_URI = "754106851545.dkr.ecr.us-west-2.amazonaws.com/pr-tensorflow-training:2.2.0-gpu-py37-cu101-ubuntu18.04-example-pr-269-2020-06-11-22-13-27"
+TEST_ECR_URI = "763104351884.dkr.ecr.us-east-1.amazonaws.com/tensorflow-training:2.2.0-gpu-py37-cu101-ubuntu18.04"
```
Comment (saimidu) [inline Scheduler/tests/query_and_cancel_test.py:27]: Tests run in us-west-2 will fail if you use the image from us-east-1. Set the region to be "us-west-2".

File: src/config/test_config.py [hunk 11]
```diff
@@ -1,6 +1,7 @@
 # Disable the test codebuild jobs to be run
-DISABLE_SANITY_TESTS = False
+DISABLE_SANITY_TESTS = True
```
Comment (YYStreet) [inline src/config/test_config.py:2]: Do revert these change before merging in

Comment (saimidu) [reply src/config/test_config.py:2]: Yes, these are meant to be reverted.

Comment (Huahan98) [reply src/config/test_config.py]: fixed
